### PR TITLE
Issue #627: Checks for doomed connections when selecting a connection.

### PR DESCRIPTION
**Pull request by Rider-Linden**
Checks for doomed connections when selecting a connection for pipelining.  Restores lost checks by Carlo Wood in url.c for old issue 1420.

### CODE REVIEW

**Comment by bagder** [comment]
Unfortunately this doesn't go through test 530 for me. Does it for you? Also, please squash and rebase the commit, it didn't merge cleanly for me.

**Comment by Rider-Linden** [comment]
No.  Sadly.  Go ahead and reject the pull request, I'm going to need to see
what's going on and fix it. There were a couple other unit tests further
down the list too where it appeared to be complaining about order of
operations returned as well (I'm not as concerned about those).

Rider

**Comment by jay** [comment]
TL;DR: For future reference it will help you to create a topic branch for your work and keep your master branch in sync with the curl github remote (usually named *upstream*).

---

In a situation like this typically you would continue working on your topic branch and commit a fix for the problem you found and then push that to your github remote (usually named *origin* -- your github [fork](https://help.github.com/articles/configuring-a-remote-for-a-fork/) of the curl repo) and that commit would appear in the pull request automatically since the topic branch is associated with the pr. Then once you are done you would add a comment that you have updated and fixed the issue and bagder or whoever is following this will see that.

See #654 for an example. Reviewer says change such and such, I change it, and the next commit appears below the comment. You can also squash the commits together in your topic branch and do a forced update of the branch, but that is not a good idea if you are working on it with anyone else.

After this issue is resolved you can reset your master branch to point to the latest commit from upstream.
~~~
git checkout master
git fetch upstream
~~~
You'll see something like `3438ce7..332414a  master     -> upstream/master`
That second one will be the latest commit from the curl repo, that's what you change master back to.
~~~
git checkout -b temp
git branch -f master 332414a
git checkout master
git branch -D temp
~~~
In the future then you could:
~~~
git checkout master
git fetch upstream && git merge --ff-only upstream/master
~~~
This way your master always has the exact same commit history as upstream/master. It seems kind of complicated but using topic branches and preserving master makes it much easier to work, branches are cheap in git.

**Comment by bagder** [comment]
Closing this for now. Open a new one if you have another take. In general HTTP pipelining is a very tricky area to get things right and I advice avoiding it and switching to HTTP/2 multiplexing instead. I would like to ultimately remove our pipelining support one day. We have several more pipelining bugs lurking in the shadows.

**Comment by patrikhson** [comment]
Hello,
I think this stuff is great, I'm not joking, take a look <http://thicriwavy.houstoncompanypartyentertainment.info/xyhnvjax>

paf@frobbit.se
